pingcap/tidb-operator
# Enable customized labels for TidbCluster pods and "use-by=end-user" services

### What problem does this PR solve?
This PR add a new feature that lets user overwrite labels and annotations for pods and services.
For more information, please refer to: Fix https://github.com/pingcap/tidb-operator/issues/2946 Fix https://github.com/pingcap/tidb-operator/issues/2345

Labels and annotations for pods:

- `TidbCluster` has `Labels` and `Annotations` fields.
- Components have their own `Labels` and `Annotations` fields, which are in `XXXSpec` objects.
- Components pods inherit TidbCluster's `Labels` and `Annotations`
- Components pods' own `Labels` and `Annotations` should override `TidbCluster` level `Labels` and `Annotations`

Labels and annotations for "use-by=end-user" Tidb service and PD service:

- `TiDBServiceSpec` and `PDSpec.Service` have `Labels` and `Annotations` fields.
- "use-by=end-user" tidb and pd service should have `Labels` and `Annotations` from `TiDBServiceSpec` and `PDSpec.Service`

### Code changes


### Tests


check the custom label/ann in tc spec will not lost in the change configuration test

### test sync ann/label from schedule to backup
1. create backupschedule with custom labels/ann
2. check generated `backup`
```yaml
apiVersion: v1
items:
- apiVersion: pingcap.com/v1alpha1
  kind: Backup
  metadata:
    annotations:
      aaaaa: ann
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"pingcap.com/v1alpha1","kind":"BackupSchedule","metadata":{"annotations":{"aaaaa":"ann"},"labels":{"alabel":"av"},"name":"basic-backup-schedule-nfs","namespace":"tidb-cluster"},"spec":{"backupTemplate":{"br":{"cluster":"basic","clusterNamespace":"tidb-cluster"},"cleanPolicy":"Delete","local":{"volume":{"name":"nfs","nfs":{"path":"/nfs","server":"192.168.0.2"}},"volumeMount":{"mountPath":"/nfs","name":"nfs"}}},"maxBackups":2,"schedule":"*/1 * * * *"}}
    creationTimestamp: "2021-05-24T03:27:13Z"
    finalizers:
    - tidb.pingcap.com/backup-protection
    generation: 2
    labels:
      alabel: av
      app.kubernetes.io/instance: basic-backup-schedule-nfs
      app.kubernetes.io/managed-by: backup-schedule-operator
      app.kubernetes.io/name: backup-schedule
```

### Related changes


### Release Notes

```release-note
- Support cusomized labels for `TidbCluster` pods
- Support cusomized labels for end-user service(TiDB service and PD service)
```
